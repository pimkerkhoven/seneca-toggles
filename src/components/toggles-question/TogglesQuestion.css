@import url('https://fonts.googleapis.com/css2?family=Mulish:wght@400;700&display=swap');

.TogglesQuestion {
    font-family: 'Mulish', sans-serif;
}

.TogglesQuestion {
    /*background: linear-gradient(180deg, #F6B868 0%, #EE6B2D 100%);*/
    display: flex;
    flex-direction: column;
    /*max-width: 900px;*/
    margin: 0 auto;
    /*justify-content: center;*/
    /*align-items: center;*/
}

.TogglesQuestion.correct {
    background: linear-gradient(180deg, #76E0C2 0%, #59CADA 100%);
}

.TogglesQuestion.partially-correct {
    background: linear-gradient(180deg, #F6B868 0%, #EE6B2D 100%);
}

.TogglesQuestion.incorrect {
    background: linear-gradient(180deg, #F1B496 0%, #EA806A 100%)
}

.TogglesQuestion h1, .TogglesQuestion h2 {
    font-style: normal;
    font-weight: 700;
    line-height: 140%;

    text-align: center;

    color: #FFFFFF;
}

.TogglesQuestion h1 {
    font-size: 40px;
}

.TogglesQuestion h2 {
    font-size: 32px;
}

@media screen and (max-width: 600px) {
    .TogglesQuestion h1, .TogglesQuestion h2 {
        line-height: 160%;
    }

    .TogglesQuestion h1 {
        font-size: 20px;
    }

    .TogglesQuestion h2 {
        font-size: 16px;
    }
}

.TogglesQuestion > * {
    margin-bottom: 24px;
}
