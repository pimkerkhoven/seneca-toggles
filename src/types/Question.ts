// TODO: add assumption to Readme
// TODO: also have assumptions in comments surrounding code
// Assume answer is included in options
interface QuestionPart {
    options: string[]
    answer: string
}

interface Question {
    title: String
    parts: QuestionPart[]
}

export default Question
