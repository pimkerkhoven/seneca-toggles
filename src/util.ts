export function shuffleArray<T>(arr: T[]) {
    arr.sort(() => Math.random() - 0.5);
}

export function randomNumber(min: number, max: number) {
    return Math.floor(Math.random() * (max - min) + min);
}