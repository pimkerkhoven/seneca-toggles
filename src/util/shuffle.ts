export function shuffleArray<T>(arr: T[]) {
    arr.sort(() => Math.random() - 0.5);
}
