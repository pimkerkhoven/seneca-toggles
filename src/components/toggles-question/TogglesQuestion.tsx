import React, {useEffect, useState} from 'react';
import Question from "../../types/Question";

interface TogglesQuestionProps {
    question: Question
}

const TogglesQuestion: React.FC<TogglesQuestionProps> = ({question: {title, parts}}) => {
    // Indicates whether the question is answered correctly (i.e. if all parts are answered correctly)
    // TODO: consistent naming of boolean variables is.../are.../has...
    const [isSolved, setIsSolved] = useState<boolean>(false)

    // TODO: maybe this works a bit nicer if answers is an object?  with a key for each part? How
    //  will typescript handle this?
    //  -> https://stackoverflow.com/questions/41045924/how-to-represent-a-variable-key-name-in-typescript-interface
    // TODO: make sure never starts in answered formation
    const [answers, setAnswers] = useState<string[]>(parts.map(part => part.options[0]))

    // TODO: include event.preventDefault
    function handleToggleAnswer(event: React.MouseEvent, partIndex: number, option: string) {
        event.preventDefault()

        if (isSolved) {
            return
        }

        setAnswers(prevState =>
            [...prevState.slice(0, partIndex), option, ...prevState.slice(partIndex + 1)])
    }

    useEffect(() => {
        // Check for each question if the current answer is the actual answer for the question
        // Combine this result with the result of the previous questions
        // TODO: maybe there is a dedicated array function for this?
        const areAllAnswersCorrect = answers.reduce((accumulator, currentAnswer, index) => {
            return accumulator && parts[index].answer === currentAnswer
        }, true)

        if (areAllAnswersCorrect) {
            setIsSolved(true)
        }
    }, [answers, parts])


    return (
        <div className="TogglesQuestion">
            <h1>{title}</h1>
            {parts.map((part, partIndex) => (
                <div key={partIndex} style={{margin: 10, border: '1px solid black'}}>
                    {part.options.map(option => {
                        const style = option === answers[partIndex] ? { fontWeight : "bold" } : {}

                        return <div key={option} style={style} onClick={event => handleToggleAnswer(event, partIndex, option)}>
                            {option}
                        </div>
                    })}
                </div>
            ))}
            <h1>{isSolved ? "The answer is correct!" : "The answer is incorrect"}</h1>
        </div>
    )
}

export default TogglesQuestion