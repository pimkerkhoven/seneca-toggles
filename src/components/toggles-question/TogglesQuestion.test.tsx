/* eslint-disable testing-library/no-node-access */

import React from 'react';
import {fireEvent, render, screen} from '@testing-library/react';
import * as util from "../../util";
import TogglesQuestion from "./TogglesQuestion";
import {Question} from "../../types/Question";


/**
 * - test snapshot of initial render
 * - test clicking answer toggles option
 * - test putting correct answer updates status text
 * - test putting correct answer locks toggles
 * - TODO: add test for testing gradient color/adding correctness class
 */

const question: Question = {
  // TODO: Where to format question? Assume is in correct format (capitalization?)
  //  Same for options
  // TODO: Assume options are all unique within part
  // TODO: State assumption of question format.
  title: "An animal cell contains:",
  parts: [
    {
      options: ["Cell wall", "Ribosomes"],
      answer: "Cell wall"
    },
    {
      options: ["Cytoplasm", "Chloroplast"],
      answer: "Cytoplasm"
    },
    {
      options: ["Partial membrane", "Impermeable membrane"],
      answer: "Impermeable membrane"
    }
  ]
}

// TODO: move tests to correct file!

beforeEach(() => {
  const shuffleArraySpy = jest.spyOn(util, 'shuffleArray')
  shuffleArraySpy.mockImplementation(_ => { })

  const randomNumberSpy = jest.spyOn(util, 'randomNumber')
  randomNumberSpy.mockImplementation((min: number, _: number) => {
    return min
  })
})

test('renders the component', () => {
  const { asFragment } = render(<TogglesQuestion question={question} />);

  expect(asFragment()).toMatchSnapshot();

  expect((screen.getByText("Cell wall").parentElement as Element).className).toBe("ToggleOption selected")
  expect((screen.getByText("Chloroplast").parentElement as Element).className).toBe("ToggleOption selected")
  expect((screen.getByText("Partial membrane").parentElement as Element).className).toBe("ToggleOption selected")
});

test('toggling an option marks it as selected', () => {
  const {asFragment} = render(<TogglesQuestion question={question} />);

  const toggle = screen.getByText("Ribosomes").parentElement as Element
  fireEvent.click(toggle)

  expect(asFragment()).toMatchSnapshot()
  expect(toggle.className).toBe( "ToggleOption selected")
  expect((screen.getByText("Cell wall").parentElement as Element).className).toBe("ToggleOption")
});

test('marking last incorrect answer correct updates status text', () => {
  const { asFragment } = render(<TogglesQuestion question={question} />);

  const statusText = screen.getByText("The answer is incorrect")

  const toggle1 = screen.getByText("Cytoplasm").parentElement as Element
  fireEvent.click(toggle1)

  // TODO: move to eslint file?
  const toggle2 = screen.getByText("Impermeable membrane").parentElement as Element
  fireEvent.click(toggle2)

  expect(asFragment()).toMatchSnapshot()
  expect(statusText.innerHTML).toBe("The answer is correct!")
});

test('marking last incorrect answer correct locks toggles', () => {
  const { asFragment } = render(<TogglesQuestion question={question} />);

  const toggle1 = screen.getByText("Cytoplasm").parentElement as Element
  fireEvent.click(toggle1)

  const toggle2 = screen.getByText("Impermeable membrane").parentElement as Element
  fireEvent.click(toggle2)

  const lockedToggle = screen.getByText("Ribosomes").parentElement as Element
  fireEvent.click(lockedToggle)

  expect(asFragment()).toMatchSnapshot()
  expect(lockedToggle.className).toBe("ToggleOption")
});


