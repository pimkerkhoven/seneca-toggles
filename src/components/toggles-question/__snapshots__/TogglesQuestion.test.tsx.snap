// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`marking last incorrect answer correct locks toggles 1`] = `
<DocumentFragment>
  <div
    class="toggles-question correct"
  >
    <h1>
      An animal cell contains:
    </h1>
    <div
      class="toggle"
    >
      <div
        class="slider"
        style="width: 50%; height: 100%; left: 0%; top: 0px;"
      />
      <div
        class="toggle-option selected"
      >
        <span>
          Cell wall
        </span>
      </div>
      <div
        class="toggle-option"
      >
        <span>
          Ribosomes
        </span>
      </div>
    </div>
    <div
      class="toggle"
    >
      <div
        class="slider"
        style="width: 50%; height: 100%; left: 0%; top: 0px;"
      />
      <div
        class="toggle-option selected"
      >
        <span>
          Cytoplasm
        </span>
      </div>
      <div
        class="toggle-option"
      >
        <span>
          Chloroplast
        </span>
      </div>
    </div>
    <div
      class="toggle"
    >
      <div
        class="slider"
        style="width: 50%; height: 100%; left: 50%; top: 0px;"
      />
      <div
        class="toggle-option"
      >
        <span>
          Partial membrane
        </span>
      </div>
      <div
        class="toggle-option selected"
      >
        <span>
          Impermeable membrane
        </span>
      </div>
    </div>
    <h2>
      The answer is correct!
    </h2>
  </div>
</DocumentFragment>
`;

exports[`marking last incorrect answer correct updates status text 1`] = `
<DocumentFragment>
  <div
    class="toggles-question correct"
  >
    <h1>
      An animal cell contains:
    </h1>
    <div
      class="toggle"
    >
      <div
        class="slider"
        style="width: 50%; height: 100%; left: 0%; top: 0px;"
      />
      <div
        class="toggle-option selected"
      >
        <span>
          Cell wall
        </span>
      </div>
      <div
        class="toggle-option"
      >
        <span>
          Ribosomes
        </span>
      </div>
    </div>
    <div
      class="toggle"
    >
      <div
        class="slider"
        style="width: 50%; height: 100%; left: 0%; top: 0px;"
      />
      <div
        class="toggle-option selected"
      >
        <span>
          Cytoplasm
        </span>
      </div>
      <div
        class="toggle-option"
      >
        <span>
          Chloroplast
        </span>
      </div>
    </div>
    <div
      class="toggle"
    >
      <div
        class="slider"
        style="width: 50%; height: 100%; left: 50%; top: 0px;"
      />
      <div
        class="toggle-option"
      >
        <span>
          Partial membrane
        </span>
      </div>
      <div
        class="toggle-option selected"
      >
        <span>
          Impermeable membrane
        </span>
      </div>
    </div>
    <h2>
      The answer is correct!
    </h2>
  </div>
</DocumentFragment>
`;

exports[`renders the component 1`] = `
<DocumentFragment>
  <div
    class="toggles-question incorrect"
  >
    <h1>
      An animal cell contains:
    </h1>
    <div
      class="toggle"
    >
      <div
        class="slider"
        style="width: 50%; height: 100%; left: 0%; top: 0px;"
      />
      <div
        class="toggle-option selected"
      >
        <span>
          Cell wall
        </span>
      </div>
      <div
        class="toggle-option"
      >
        <span>
          Ribosomes
        </span>
      </div>
    </div>
    <div
      class="toggle"
    >
      <div
        class="slider"
        style="width: 50%; height: 100%; left: 50%; top: 0px;"
      />
      <div
        class="toggle-option"
      >
        <span>
          Cytoplasm
        </span>
      </div>
      <div
        class="toggle-option selected"
      >
        <span>
          Chloroplast
        </span>
      </div>
    </div>
    <div
      class="toggle"
    >
      <div
        class="slider"
        style="width: 50%; height: 100%; left: 0%; top: 0px;"
      />
      <div
        class="toggle-option selected"
      >
        <span>
          Partial membrane
        </span>
      </div>
      <div
        class="toggle-option"
      >
        <span>
          Impermeable membrane
        </span>
      </div>
    </div>
    <h2>
      The answer is incorrect
    </h2>
  </div>
</DocumentFragment>
`;

exports[`toggling an option marks it as selected 1`] = `
<DocumentFragment>
  <div
    class="toggles-question incorrect"
  >
    <h1>
      An animal cell contains:
    </h1>
    <div
      class="toggle"
    >
      <div
        class="slider"
        style="width: 50%; height: 100%; left: 50%; top: 0px;"
      />
      <div
        class="toggle-option"
      >
        <span>
          Cell wall
        </span>
      </div>
      <div
        class="toggle-option selected"
      >
        <span>
          Ribosomes
        </span>
      </div>
    </div>
    <div
      class="toggle"
    >
      <div
        class="slider"
        style="width: 50%; height: 100%; left: 50%; top: 0px;"
      />
      <div
        class="toggle-option"
      >
        <span>
          Cytoplasm
        </span>
      </div>
      <div
        class="toggle-option selected"
      >
        <span>
          Chloroplast
        </span>
      </div>
    </div>
    <div
      class="toggle"
    >
      <div
        class="slider"
        style="width: 50%; height: 100%; left: 0%; top: 0px;"
      />
      <div
        class="toggle-option selected"
      >
        <span>
          Partial membrane
        </span>
      </div>
      <div
        class="toggle-option"
      >
        <span>
          Impermeable membrane
        </span>
      </div>
    </div>
    <h2>
      The answer is incorrect
    </h2>
  </div>
</DocumentFragment>
`;
